contracts/integration/MoC/MocIntegration.sol

```diff
@@ -98,6 +113,47 @@ contract MocIntegration is OwnableUpgradeable, ERC1967UpgradeUpgradeable {
         emit GetDocFromDllrAndRedeemRBTC(msg.sender, _dllrAmount, rbtcAmount);
     }
 
+    /**
+     * @notice how getDocFromDllrAndRedeemRBTC function works:
+     * -------------------------------------------------------------------------------------------
+     * |               Mynt                         |                Money On Chain              |
+     * -------------------------------------------------------------------------------------------
+     * | get DLLR (EIP-2612) -> convert DLLR to DoC | -> get RBTC from DoC -> send RBTC to user  |
+     * -------------------------------------------------------------------------------------------
+     *
+     * @param permit permit data, in form of PermitTransferFrom struct.
+     * @param signature of the permit data.
+     */
+    function getDocFromDllrAndRedeemRbtcWithPermit2(
+        ISignatureTransfer.PermitTransferFrom memory permit,
+        bytes memory signature
+    ) external {
+        address thisAddress = address(this);
+        uint256 _dllrAmount = permit.permitted.amount;
+
+        ISignatureTransfer.SignatureTransferDetails
+            memory transferDetails = _generateTransferDetails(thisAddress, _dllrAmount);
+
+        permit2.permitTransferFrom(permit, transferDetails, msg.sender, signature);
+
+        _useNonce(msg.sender);
```

Comment:
I don't understand how this code validates the nonces. Where does it check if a nonce has already been used?

Reply:
The nonce is actually just a unique number that is included when user sign the permitTransferFrom message, it's not even necessarily added to the contract (as long FE can detect the unique number).
But since FE can't do it, that's the reason why we add nonce in the integrated contract
so everytime FE call the function, needs to get the nonce from the integrated contract first
but in case they send the same nonce, it will be rejected by permit2 contract

Reply:
the nonce is actually just a unique number (which is included as part of `permitTransferFrom` message), it's not even necessarily added to the contract (as long FE can detect the unique number).
But since FE can't do it, thats the reason why we add nonce in the integrated contract
so everytime FE call the function, needs to get the nonce from the integrated contract first
but in case they send the same nonce, it will be rejected by permit2 contract